mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,284 @@
+            _ => {
+                self.pending = None;
+            }
+        }
+    }
+}
+
+/// Marker type for an IO implementation that doesn't implement
+/// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+pub enum AnyAsyncIo {}
+#[cfg(feature = "std")]
+/// Marker type for an IO implementation that implements both
+/// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+pub enum AsyncStream {}
+
+#[async_trait]
+pub trait AsyncIo {
+    async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
+    async fn send(&mut self, buf: &[u8]) -> Result<usize>;
+}
+
+impl<'a, 'b, 'c, IO: AsyncIo> IoCallback<AnyAsyncIo> for (&'a mut TaskContext<'b>, &'c mut IO) {
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        match self.1.recv(buf).as_mut().poll(self.0) {
+            Poll::Ready(Ok(n)) => Ok(n),
+            Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+            Poll::Pending => Err(Error::SslWantRead),
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        match self.1.send(buf).as_mut().poll(self.0) {
+            Poll::Ready(Ok(n)) => Ok(n),
+            Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+            Poll::Pending => Err(Error::SslWantWrite),
+        }
+    }
+}
+
+impl<'a, 'b, 'c, IO: AsyncRead + AsyncWrite + std::marker::Unpin + 'static> IoCallback<AsyncStream>
+    for (&'a mut TaskContext<'b>, &'c mut IO)
+{
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut buf = ReadBuf::new(buf);
+        let io = Pin::new(&mut self.1);
+        match io.poll_read(self.0, &mut buf) {
+            Poll::Ready(Ok(_)) => Ok(buf.filled().len()),
+            Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+            Poll::Pending => Err(Error::SslWantRead),
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        let io = Pin::new(&mut self.1);
+        match io.poll_write(self.0, buf) {
+            Poll::Ready(Ok(n)) => Ok(n),
+            Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+            Poll::Pending => Err(Error::SslWantWrite),
+        }
+    }
+}
+
+struct HandshakeFuture<'a, T>(&'a mut Context<T>);
+
+impl<T> Future for HandshakeFuture<'_, T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    type Output = Result<()>;
+    fn poll(mut self: Pin<&mut Self>, ctx: &mut TaskContext) -> std::task::Poll<Self::Output> {
+        self.0
+            .with_bio_async(ctx, |ssl_ctx| match ssl_ctx.handshake() {
+                Err(Error::SslWantRead) | Err(Error::SslWantWrite) => Poll::Pending,
+                Err(e) => Poll::Ready(Err(e)),
+                Ok(()) => Poll::Ready(Ok(())),
+            })
+            .unwrap_or(Poll::Ready(Err(Error::NetSendFailed)))
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite + 'static> Context<T> {
+    pub async fn accept_async<IoType>(config: Arc<Config>, io: T, hostname: Option<&str>) -> IoResult<Context<T>>
```

Comment:
We don't have any special constructors anymore, I think we can get rid of this now. Caller can do this.

Reply:
Yeah. Removed